adal/log.py

```diff
@@ -93,16 +95,63 @@ def _log_message(self, msg, log_stack_trace=None):
         return formatted
 
     def warn(self, msg, *args, **kwargs):
+        """
+        The recommended way to call this function with variable content,
+        is to use the `warn("hello %(name)s", {"name": "John Doe"}` form,
+        so that this method will scrub pii value when needed.
+        """
+        if len(args) == 1 and isinstance(args[0], dict) and not self.log_context.get('enable_pii'):
+            args = (scrub_pii(args[0]),)
         log_stack_trace = kwargs.pop('log_stack_trace', None)
         msg = self._log_message(msg, log_stack_trace)
         self._logging.warning(msg, *args, **kwargs)
 
     def info(self, msg, *args, **kwargs):
+        if len(args) == 1 and isinstance(args[0], dict) and not self.log_context.get('enable_pii'):
+            args = (scrub_pii(args[0]),)
         log_stack_trace = kwargs.pop('log_stack_trace', None)
         msg = self._log_message(msg, log_stack_trace)
         self._logging.info(msg, *args, **kwargs)
 
     def debug(self, msg, *args, **kwargs):
+        if len(args) == 1 and isinstance(args[0], dict) and not self.log_context.get('enable_pii'):
+            args = (scrub_pii(args[0]),)
         log_stack_trace = kwargs.pop('log_stack_trace', None)
         msg = self._log_message(msg, log_stack_trace)
         self._logging.debug(msg, *args, **kwargs)
+
+    def exception(self, msg, *args, **kwargs):
+        if len(args) == 1 and isinstance(args[0], dict) and not self.log_context.get('enable_pii'):
+            args = (scrub_pii(args[0]),)
+        msg = self._log_message(msg)
+        self._logging.exception(msg, *args, **kwargs)

```

Comment:
I don't have particular opinions on this change, but just in case it helps though, that this will log the stuff to STDERR, instead of STDOUT, hence the change you made in other places to switching to this new method will be a breaking change.

Reply:
Interestingly, I also thought exception logs would go into STDERR, while most other less severe logs would go to STDOUT. Turns out, that is not the case. Quoting from the [Advanced Logging Tutorial](https://docs.python.org/2/howto/logging.html#logging-advanced-tutorial), and I've also done an experiment to confirm that:

> If you call the functions debug(), info(), warning(), error() and critical(), they will check to see if no destination is set; and if one is not set, they will set a destination of the console (sys.stderr)

Technically, this PR does switch some previous INFO or WARNING logs into EXCEPTION (which is equivalent to ERROR), though. This kind of adjustment shouldn't bring any breaking change. Please let me know if I missed anything.